File: library/src/test/java/io/ashdavies/rx/rxfirebase/ValueEventSingleTest.java
package io.ashdavies.rx.rxfirebase;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseException;
import com.google.firebase.database.Query;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import io.reactivex.SingleEmitter;

import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class ValueEventSingleTest {

  private ValueEventSingle single;

  @Mock SingleEmitter<DataSnapshot> emitter;

  @Before
  public void setUp() throws Exception {
    single = new ValueEventSingle(emitter);
  }

  @Test
  public void shouldEmitDataSnapshot() throws Exception {
    DataSnapshot snapshot = mock(DataSnapshot.class);

    single.onDataChange(snapshot);

    verify(emitter).onSuccess(snapshot);
  }

  @Test
  public void shouldEmitException() throws Exception {
    DatabaseError error = mock(DatabaseError.class);

    DatabaseException exception = mock(DatabaseException.class);
    given(error.toException()).willReturn(exception);

    single.onCancelled(error);

    verify(emitter).onError(exception);
  }

  @Test
  public void shouldRemoveEventListener() throws Exception {
    Query query = mock(Query.class);

    single.cancellable(query).cancel();

    verify(query).removeEventListener(single);
  }
}

File: src/test/kotlin/io/ashdavies/rx/rxfirebase/ValueEventSingleTest.kt
package io.ashdavies.rx.rxfirebase

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseException
import com.nhaarman.mockito_kotlin.isA
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.then
import io.reactivex.SingleEmitter
import org.junit.jupiter.api.Test

internal class ValueEventSingleTest {

  private val emitter: SingleEmitter<DataSnapshot> = mock()
  private val snapshot: DataSnapshot = mock()

  private val listener = ValueEventSingle(emitter)

  @Test
  fun `should emit snapshot`() {
    listener.onDataChange(snapshot)

    then(emitter)
        .should()
        .onSuccess(snapshot)
  }

  @Test
  fun `should emit exception`() {
    listener.onCancelled(DatabaseError.fromCode(418))

    then(emitter)
        .should()
        .onError(isA<DatabaseException>())
  }

}
